provd/internal/services/user/user.go

```
@@ -186,51 +201,83 @@ func (s *Service) ValidateUsername(ctx context.Context, req *pb.ValidateUsername
 	}
 
 	// Check if username is already in use
-	err = s.Accounts.Call(consts.DbusAccountsPrefix+".FindUserByName", 0, username).Err
-	if err != nil {
-		var dbusError dbus.Error
-		if errors.As(err, &dbusError) {
-			if dbusError.Name == consts.DbusAccountsPrefix+".Error.Failed" {
-				// User not found
-				return &pb.ValidateUsernameResponse{UsernameValidation: pb.UsernameValidation_OK}, nil
-			}
-			// Handle other dbus errors
-			return nil, status.Errorf(codes.Internal, "dbus error: %v", dbusError)
-		}
-		// Unknown error
+	err = s.accounts.Call(consts.DbusAccountsPrefix+".FindUserByName", 0, username).Err
+
+	// User found
+	if err == nil {
+		return &pb.ValidateUsernameResponse{UsernameValidation: pb.UsernameValidation_ALREADY_IN_USE}, nil
+	}
+
+	// Inspect the error we got
+	var dbusError dbus.Error
+	isDbusError := errors.As(err, &dbusError)
+
+	// Unknown error
+	if !isDbusError {
 		return nil, status.Errorf(codes.Internal, "unknown error: %v", err)
 	}
 
-	// User found
-	return &pb.ValidateUsernameResponse{UsernameValidation: pb.UsernameValidation_ALREADY_IN_USE}, nil
+	// Non "user not found" account service error
+	errBody, ok := dbusError.Body[0].(string)
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "unknown error: %v", err)
+	}
+
+	if dbusError.Name != consts.DbusAccountsPrefix+".Error.Failed" || !strings.Contains(errBody, username) {
+		return nil, status.Errorf(codes.Internal, "dbus error: %v", dbusError)
+	}
+
+	// We have "user not found"
+	return &pb.ValidateUsernameResponse{UsernameValidation: pb.UsernameValidation_OK}, nil
+}
+
+// Function to scan a file for the username.
+func scanFile(file *os.File, username string) (bool, error) {
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+		if strings.HasPrefix(line, "#") || line == "" {
+			continue
+		}
+		// Extract the username from the line
+		fields := strings.SplitN(line, ":", 2)
```

I would error out if the file does not have the expected format, So failing early, continuing if not found, and the "happy path" is one tab more left, with the return.